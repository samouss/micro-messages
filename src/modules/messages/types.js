// @flow

export type MessageId = string;
export type MessageBody = string;

export type Visibility = 'public' | 'private';

export type VisibilityFilter = Visibility | 'all';

export type Message = {
  id: MessageId,
  date: string,
  body: MessageBody,
  visibility: Visibility,
};

export type MessageInput = {
  body: MessageBody,
  visibility: Visibility,
};

export type ByIdState = { [id: MessageId]: Message };
export type IdsState = Array<MessageId>;

export type MessageState = {
  byId: ByIdState,
  ids: IdsState,
};

export type FetchMessagesRequestAction = { type: 'FETCH_MESSAGES_REQUEST' };
export type FetchMessagesSuccessAction = { type: 'FETCH_MESSAGES_SUCCESS', messages: Array<Message> };
export type FetchMessagesFailureAction = { type: 'FETCH_MESSAGES_FAILURE' };

export type PostMessageRequestAction = { type: 'POST_MESSAGE_REQUEST', message: Message };
export type PostMessageSuccessAction = { type: 'POST_MESSAGE_SUCCESS', message: Message };
export type PostMessageFailureAction = { type: 'POST_MESSAGE_FAILURE' };

export type MessageAction =
  | FetchMessagesRequestAction
  | FetchMessagesSuccessAction
  | FetchMessagesFailureAction

  | PostMessageRequestAction
  | PostMessageSuccessAction
  | PostMessageFailureAction;
