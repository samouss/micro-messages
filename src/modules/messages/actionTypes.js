// @flow

export const FETCH_MESSAGES_REQUEST = 'FETCH_MESSAGES_REQUEST';
export const FETCH_MESSAGES_SUCCESS = 'FETCH_MESSAGES_SUCCESS';
export const FETCH_MESSAGES_FAILURE = 'FETCH_MESSAGES_FAILURE';

export const POST_MESSAGE_REQUEST = 'POST_MESSAGE_REQUEST';
export const POST_MESSAGE_SUCCESS = 'POST_MESSAGE_SUCCESS';
export const POST_MESSAGE_FAILURE = 'POST_MESSAGE_FAILURE';
