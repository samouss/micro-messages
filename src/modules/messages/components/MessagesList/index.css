.MessagesList {
  padding-top: 35px;
}
