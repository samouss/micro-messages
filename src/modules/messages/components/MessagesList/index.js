// @flow

import React from 'react';
import MessageItem from 'modules/messages/components/MessageItem';
import './index.css';

import type { Message } from 'modules/messages/types';

export type Props = {
  messages: Array<Message>,
};

const MessagesList = ({ messages }: Props) => (
  <ul styleName="MessagesList">
    {messages.map(message => (
      <MessageItem
        key={message.id}
        message={message}
      />
    ))}
  </ul>
);

export default MessagesList;
