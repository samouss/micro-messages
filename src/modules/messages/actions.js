// @flow

import { fetchMessages as fetchMessagesFromApi } from 'core/api';
import type { Dispatch } from 'store/types';
import * as ACTION_TYPES from './actionTypes';
import type { Message, MessageAction } from './types';

const fetchMessagesRequest = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_REQUEST,
});

const fetchMessagesSuccess = (messages: Array<Message>): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_SUCCESS,
  messages,
});

const fetchMessagesFailure = (): MessageAction => ({
  type: ACTION_TYPES.FETCH_MESSAGES_FAILURE,
});

export const fetchMessages = () => (dispatch: Dispatch) => {
  dispatch(fetchMessagesRequest());

  return fetchMessagesFromApi().then(res => {
    dispatch(fetchMessagesSuccess(res));
  }).catch(() => {
    dispatch(fetchMessagesFailure());
  });
};
