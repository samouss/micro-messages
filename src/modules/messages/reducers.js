// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { State, Action } from 'store/types';
import type { MessageId, ByIdState, IdsState } from './types';

const byIdInitialState: ByIdState = {};
const byId = (state = byIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const idsInitialState: IdsState = [];
const ids = (state = idsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  byId,
  ids,
});

const getState = (state: State) => state.messages;
const getMessageIds = (state: State) => getState(state).ids;

export const getMessage = (state: State, id: MessageId) => {
  return getState(state).byId[id];
};

export const getMessages = (state: State) => {
  return getMessageIds(state).map(id => getMessage(state, id));
};
