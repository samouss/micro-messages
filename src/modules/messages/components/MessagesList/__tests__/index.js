// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockMessage } from 'test/messages';
import MessagesList from '../index';

import type { Props } from '../index';

describe('<MessagesList />', () => {
  it('expect to render the component correctly', () => {
    const props: Props = {
      messages: [
        createMockMessage('ID1234'),
        createMockMessage('ID5678'),
        createMockMessage('ID9101'),
      ],
    };

    const component = shallow(
      <MessagesList
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });
});
