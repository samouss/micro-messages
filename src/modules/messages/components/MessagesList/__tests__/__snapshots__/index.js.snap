// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessagesList /> expect to render the component correctly 1`] = `
<ul>
  <MessageItem
    message={
      Object {
        "body": "BODY",
        "date": "SOME_DATE",
        "id": "ID1234",
        "visibility": "private",
      }
    }
  />
  <MessageItem
    message={
      Object {
        "body": "BODY",
        "date": "SOME_DATE",
        "id": "ID5678",
        "visibility": "private",
      }
    }
  />
  <MessageItem
    message={
      Object {
        "body": "BODY",
        "date": "SOME_DATE",
        "id": "ID9101",
        "visibility": "private",
      }
    }
  />
</ul>
`;
